### ISSUE TITLE
[feature request] Make `torch.gather` broadcastable.

### ISSUE THREAD

**Issue by zasdfgbnm** [pytorch/pytorch#9407]
Currently, `torch.gather` does not broadcast. For example:
```python
t = torch.tensor([[1,2],[1,2]])
torch.gather(t, 1, torch.tensor([[0,1,0],[0,1,0]]))
```
gives
```
tensor([[ 1,  2,  1],
        [ 1,  2,  1]])
```

But

```python
t = torch.tensor([[1,2],[1,2]])
torch.gather(t, 1, torch.tensor([[0,1,0]]))
```
and
```python
t = torch.tensor([[1,2]])
torch.gather(t, 1, torch.tensor([[0,1,0],[0,1,0]]))
```
gives runtime errors.

It would be more convenient to allow broadcasting on every dimension except the dim specified as the second argument in gather, so that the above two codes can give the same result as the first code.

Does this proposal sound good? If yes, I can work on it.

**Comment by gchanan** [comment]
I think this would be useful.  You may want to look at my broadcasting notes: https://github.com/gchanan/pytorch/wiki/Broadcasting-Notes, which describes what to do with `torch.gather`.

**Comment by zasdfgbnm** [comment]
@gchanan 
I'm not sure if I understand the explanation at [gather-explanation](https://github.com/gchanan/pytorch/wiki/Broadcasting-Notes#gather-explanation):

> i <= index.dim() < n,  Note this doesn't come up for numpy.

Shouldn't this always be the case for numpy? Since numpy always has `dim=0`. Also, shouldn't the examples in `i > index.dim()` case belong to the `i <= index.dim() < n` case? Also, is there any difference between `i <= index.dim() < n` and `index.dim() < i < n`, shouldn't we always prepend index to make it the same dim as tensor? Also, should there be a `n > index.dim()` case? In this case we prepend dims to the `tensor` tensor, rather than to `index` tensor.

**Comment by nctubuddy** [comment]
Is this feature still being worked on? It seems all related PRs have been closed.

**Comment by lqf96** [comment]
Note that [`tf.gather`](https://www.tensorflow.org/api_docs/python/tf/gather) and [`tf.gather_nd`](https://www.tensorflow.org/api_docs/python/tf/gather_nd) both support it and it will make indexing much easier in certain cases.

**Comment by okhat** [comment]
Any updates on this? It would be pretty useful!

**Comment by gauenk** [comment]
Agreed. I need this frequently in my code!

**Comment by 99991** [comment]
In many cases, you can achieve the same effect as `torch.gather` with indexing, because indexing supports broadcasting and you can use multiple indexing tensors at once.

Here is an example for gathering over a 4-dimensional tensor while broadcasting over one dimension:

```python
import torch

def gather(inputs, dim, indices):
    dim_indices = []
    for i, size in enumerate(inputs.shape):
        shape = [1] * inputs.ndim
        shape[i] = -1
        dim_indices.append(torch.arange(size).reshape(shape))
    dim_indices[dim] = indices

    return inputs[tuple(dim_indices)]

def test():
    inputs = torch.rand(5, 6, 7, 8)

    for dim in range(inputs.ndim):
        shape = list(inputs.shape)

        if dim > 0:
            shape[dim - 1] = 1

        indices = torch.randint(0, inputs.shape[dim], size=shape)

        shape = [1] * inputs.ndim

        if dim > 0:
            shape[dim - 1] = inputs.shape[dim - 1]

        broadcast_indices = indices.repeat(shape)

        expected_outputs = torch.gather(inputs, dim, broadcast_indices)

        if dim > 0:
            print(f"Gather with {indices.shape} broadcasting over dim {dim - 1}.")

        outputs = gather(inputs, dim, indices)

        assert outputs.shape == expected_outputs.shape
        assert torch.allclose(outputs, expected_outputs)

test()
```
